core/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/audio/environment/AudioEnvironment.java

```diff
@@ -0,0 +1,256 @@
+	 * @param off
+	 *            the start offset in the data
+	 * @param len
+	 *            the number of bytes to write
+	 */
+	protected abstract void writeToOutput(byte[] buffer, int off, int len);
+
+	/**
+	 * @return The {@link AudioFormat} format for outputting. The {@link AudioOutput} given to
+	 *         {@link #playAudio(AudioOutput, de.unistuttgart.iaas.amyassist.amy.core.audio.AudioManager.OutputBehavior)}
+	 *         must have this format.
+	 */
+	public abstract AudioFormat getOutputFormat();
+
+	/**
+	 * @return The {@link AudioFormat} format for inputting. The {@link AudioInputStream} returned by
+	 *         {@link #getAudioInputStream()} will have this format.
+	 */
+	public abstract AudioFormat getInputFormat();
+
+	/**
+	 * Get the audio environment identifier of this audio environment.
+	 * 
+	 * @return The identifier of this audio environment
+	 */
+	public abstract UUID getAudioEnvironmentIdentifier();
+
+	/**
+	 * Stops the current output.
+	 */
+	public void stopOutput() {
+		this.outputWorker.cancel(true);
+	}
+
+	/**
+	 * Plays the given audio output. When there is already a output stream being played it behaves like defined.
+	 * 
+	 * @param audioToPlay
+	 *            The audio output to play. This must have the format returned by {@link #getOutputFormat()}.
+	 * @param behavior
+	 *            How to behave if there is an output already
+	 * @throws IllegalArgumentException
+	 *             When the audioToPlay has the wrong format.
+	 * @throws IllegalStateException
+	 *             When the behavior is unknown
+	 */
+	public void playAudio(AudioOutput audioToPlay, AudioManager.OutputBehavior behavior) {
+
+		if (!audioToPlay.getFormat().matches(getOutputFormat()))
+			throw new IllegalArgumentException("AudioToPlay has the wrong AudioFormat.");
+
+		synchronized (this.outputQueue) {
+			switch (behavior) {
+			case INTERRUPT_ALL:
+				this.outputQueue.clear();
+				this.outputWorker.cancel(true);
+				this.outputQueue.addFirst(audioToPlay);
+				break;
+			case INTERRUPT_CURRENT:
+				this.outputWorker.cancel(true);
+				this.outputQueue.addFirst(audioToPlay);
+				break;
+			case QUEUE:
+				this.outputQueue.add(audioToPlay);
+				break;
+			case QUEUE_PRIORITY:
+				this.outputQueue.addFirst(audioToPlay);
+				break;
+			case SUSPEND:
+				this.outputWorker.cancel(false);
+				this.outputQueue.addFirst(audioToPlay);
+				break;
+			default:
+				throw new IllegalStateException("Unknown behavior");
+			}
+		}
+	}
+
+	/**
+	 * Get's a {@link AudioInputStream} from this audio environment. This stream will contain the audio input (like
+	 * microphone data) of this environment.
+	 * 
+	 * This stream needs to be read from fast enough. Otherwise this stream will end.
+	 * 
+	 * When this stream is no longer needed, it should be closed.
+	 * 
+	 * @return The stream.
+	 */
+	public AudioInputStream getAudioInputStream() {
+		QueuedInputStream qis = new QueuedInputStream();
+		this.inputStreams.add(qis);
```

Comment:
how are inputStreams removed form this list when they are closed?

Reply:
The input worker does that.